lib/AST/ASTImporter.cpp

```diff
@@ -908,8 +908,27 @@ bool ASTNodeImporter::ImportDeclParts(NamedDecl *D, DeclContext *&DC,
                                       DeclarationName &Name, 
                                       NamedDecl *&ToD,
                                       SourceLocation &Loc) {
+  // Check if RecordDecl is in FunctionDecl parameters to avoid infinite loop.
+  DeclContext *OrigDC = D->getDeclContext();
+  if(isa<RecordDecl>(D) && OrigDC->isFunctionOrMethod()) {
+    FunctionDecl *FunDecl = dyn_cast<FunctionDecl>(OrigDC);
```

Comment:
Can this cast fail after `isFunctionOrMethod` is true?

Reply:
Well, isFunctionOrMethod allows some other types. I was scared of ObjectiveC stuff, so it seems that FuncDecl is narrower than what is reported as true by this function. Maybe I'm wrong. The main reason using that was that it is faster (using type bits). But I may revert to not use this func and only do FuncDecl-based detection.

Reply:
Having an Obj-C test-case would help us figure this out.

Reply:
Unfortunately, current unittest framework cannot handle Objective-C. In spite of giving LANGOBJCXX it starts from a .cc file and handles the code as C. Apparently, noboy wrote Objective-C unittests yet for the importer :)